libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,966 @@
+		_params.position.line = _json["position"]["line"].asInt() + 1;
+		_params.position.column = _json["position"]["character"].asInt() + 1;
+	}
+
+	DocumentPosition extractDocumentPosition(Json::Value const& _json)
+	{
+		DocumentPosition dpos{};
+		dpos.path = extractPathFromFileURI(_json["textDocument"]["uri"].asString()).value();
+		dpos.position.line = _json["position"]["line"].asInt() + 1;
+		dpos.position.column = _json["position"]["character"].asInt() + 1;
+		return dpos;
+	}
+
+	Json::Value toJsonRange(SourceLocation const& _location)
+	{
+		solAssert(_location.source.get() != nullptr, "");
+
+		Json::Value json;
+
+		auto const [startLine, startColumn] = _location.source->translatePositionToLineColumn(_location.start);
+		json["start"]["line"] = max(startLine - 1, 0);
+		json["start"]["character"] = max(startColumn - 1, 0);
+
+		auto const [endLine, endColumn] = _location.source->translatePositionToLineColumn(_location.end);
+		json["end"]["line"] = max(endLine - 1, 0);
+		json["end"]["character"] = max(endColumn - 1, 0);
+
+		return json;
+	}
+
+	Json::Value toJson(SourceLocation const& _location)
+	{
+		solAssert(_location.source.get() != nullptr, "");
+		Json::Value item = Json::objectValue;
+		item["uri"] = toFileURI(_location.source->name());
+		item["range"] = toJsonRange(_location);
+		return item;
+	}
+
+	Json::Value toJson(LineColumn _pos)
+	{
+		Json::Value json = Json::objectValue;
+		json["line"] = max(_pos.line - 1, 0);
+		json["character"] = max(_pos.column - 1, 0);
+
+		return json;
+	}
+} // }}} end helpers
+
+LanguageServer::LanguageServer(Logger _logger, istream& _in, ostream& _out):
+	m_client{make_unique<JSONTransport>(_in, _out, _logger)},
+	m_handlers{
+		{"cancelRequest", [](auto, auto) {/*don't do anything for now, as we're synchronous*/}},
+		{"$/cancelRequest", [](auto, auto) {/*don't do anything for now, as we're synchronous*/}},
+		{"initialize", bind(&LanguageServer::handle_initialize, this, _1, _2)},
+		{"initialized", {} },
+		{"shutdown", [this](auto, auto) { m_shutdownRequested = true; }},
+		{"workspace/didChangeConfiguration", bind(&LanguageServer::handle_workspace_didChangeConfiguration, this, _1, _2)},
+		{"textDocument/didOpen", bind(&LanguageServer::handle_textDocument_didOpen, this, _1, _2)},
+		{"textDocument/didChange", bind(&LanguageServer::handle_textDocument_didChange, this, _1, _2)},
+		{"textDocument/didClose", [this](auto, Json::Value const& _args) {
+			documentClosed(extractPathFromFileURI(_args["textDocument"]["uri"].asString()).value());
+		}},
+		{"textDocument/definition", bind(&LanguageServer::handle_textDocument_definition, this, _1, _2)},
+		{"textDocument/documentHighlight", bind(&LanguageServer::handle_textDocument_highlight, this, _1, _2)},
+		{"textDocument/references", bind(&LanguageServer::handle_textDocument_references, this, _1, _2)},
+	},
+	m_logger{std::move(_logger)},
+	m_vfs()
+{
+}
+
+void LanguageServer::changeConfiguration(Json::Value const& _settings)
```

Comment:
Why not just use the standard-json settings sub-object?

Reply:
I may not get what you mean. This function `changeConfiguration` is called in two locations, once it's json object name is called "settings" and in the other one it's called "initializationOptions". So I created this function as they both mean the same (but called from a different rpc call).